core/replica.go

```diff
@@ -51,6 +52,33 @@ type Replica struct {
 	handleStream messageStreamHandler
 }
 
+// GetReplica creates a new instance of replica node
+func GetReplica(id uint32, configer api.Configer, stack Stack, handler api.ProtocolHandler, logger *logging.Logger) (*Replica, error) {
+	n := configer.N()
+	f := configer.F()
+
+	if n < 2*f+1 {
+		return nil, fmt.Errorf("%d nodes is not enough to tolerate %d faulty", n, f)
+	}
+
+	replica := &Replica{
+		id: id,
+		n:  n,
+
+		stack: stack,
+
+		log: messagelog.New(),
+
+		handleStream: makeMessageStreamHandler(handler, logger),
```

Comment:
Basically, this would allow to reuse code from `makeMessageStreamHandler`, but that function depends on message definitions which are MinBFT-specific. Perhaps, if we make the message definition generic enough, that function would become quite trivial.

Reply:
I think the workflow implemented by the function is both generic and useful enough, maybe it suffices to adapt it to work with `interface {}` instead of `*messages.Message`?

Reply:
I suggest you try and see if it works.

Reply:
This might even be the most critical piece of feedback to start addressing.

Reply:
The solution I proposed seems to work. Please check commit 40ba273008f3263bd43686648006b1fe82df0fdb, which provides additional methods to the ProtocolHandler interface to serialize/deserialize messages. Don't mind the absence of documentation for the Wrap method, it is dealt with in the subsequent commit.

Reply:
It seems to be a bit too much fuss around trying to reuse such a simple function as `makeMessageStreamHandler`. The external API tries to be generic enough to handle most of generic stuff. Maybe try to push some more generic code behind the external API? Nevertheless, if we are to go with the original proposal, it might make sense to start with some refactoring of the original code. 

Firstly, we could define a high-level abstract message type (interface) that knows how to Marshal/Unmarshal itself, as well as to give its human-readable text representation. Then this would be the job of the message implementation, not of `ProtocolHandler`. To support unmarshalling, `makeMessageStreamHandler` would receive as a parameter a mechanism to create empty messages (message factory). Empty messages would be initialized by unmarshalling the received bytes.

Secondly, we could allow for any number of reply messages, not just only one. Think of reversing the direction of peer message flow. A replica would receive from its peer just a single Hello message, then reply with a stream of messages from its message log.

Reply:
After checking out your pull request, I think an even simpler approach would be to just use a `Replica` interface. So instead of using `minbft.New`, developers that just want to re-use the connectivity could just make their own implementation of that interface.

Reply:
So what's your plan regarding this PR?

Reply:
I would suggest your PR to take priority over this one. Once it is merged with the master branch, l will pull the commits and work with what you added. If I find that none of what I proposed in ths PR is useful after the update, then this should simply be closed.